metaflow/plugins/kfp/kfp.py

```diff
@@ -212,80 +222,91 @@ def _command(
+            run_id="{run_id}",
+            step_name=step_name,
+            task_id="${{task_id_format%.}}",
+            retry_count=f"`python -c '{retry_count_python}'`",
+            datastore_type="s3",
+            datastore_root="$METAFLOW_DATASTORE_SYSROOT_S3",
+            stdout_path=STDOUT_PATH,
+            stderr_path=STDERR_PATH,
+        )
+
+        if self.s3_code_package:
+            init_cmds = environment.get_package_commands(
+                code_package_url, is_kfp_plugin=True
+            )
+        else:
+            init_cmds = [
+                "cd " + str(Path(inspect.getabsfile(self.flow.__class__)).parent)
+            ]
+
+        init_expr = " && ".join(init_cmds)
+
+        step_cmds = []
+        step_cmds.extend(environment.bootstrap_commands(step_name))
+        step_cmds.append("echo 'Task is starting.'")
+        step_cmds.extend(step_cli)
+
+        step_expr = bash_capture_logs(" && ".join(step_cmds))
+
         if "volume" in resource_requirements:
             volume_dir = resource_requirements["volume_dir"]
             clean_volume = f"rm -rf {os.path.join(volume_dir, '*')}"
         else:
+            # the `true` command is to make sure that the generated command
+            # plays well with docker containers which have entrypoint set as
+            # eval $@
             clean_volume = "true"
 
-        # We capture the exit code at two places:
-        # Once after the subshell/redirection commands, and once after the saving logs
-        # command. If either of these exit codes are not 0, we exit with the nonzero
-        # exit code manually because combining bash commands with ';' always results
-        # in an exit code of 0, whether or not certain commands failed.
-        return (
-            f"{set_debug_command}; {clean_volume}; eval `python -c '{retry_count_python}'`; "
-            f"({subshell_commands}) {redirection_commands}; export exit_code_1=$?; "
-            f"{cp_logs_cmd}; export exit_code_2=$?; "
-            f'if [ "$exit_code_1" -ne 0 ]; then exit $exit_code_1; else exit $exit_code_2; fi'
+        # construct an entry point that
+        # 1) Clean attached volume if any
+        # 2) Initializes the mflog environment (mflog_expr)
+        # 3) Bootstraps a metaflow environment (init_expr)
+        # 4) Executes a task (step_expr)
+        cmd_str = (
+            f"{clean_volume} "
+            f"&& mkdir -p {LOGS_DIR} && {task_id_format_cmd} && {mflog_expr} "
+            f"&& {init_expr} "
+            f"&& {step_expr};"
         )
 
+        # after the task has finished, we save its exit code (fail/success)
+        # and persist the final logs. The whole entrypoint should exit
+        # with the exit code (c) of the task.
+        #
+        # Note that if step_expr OOMs, this tail expression is never executed.
+        # We lose the last logs in this scenario (although they are visible
+        # still through AWS CloudWatch console).
```

Comment:
Is this valid?

`
(although they are visible
        # still through AWS CloudWatch console)
`

Reply:
> We lose the last logs in this scenario
What happens if the Flow crashes in the middle of execution, are logs still copied?  They were copied in our previous implementation.
This is a good test to have.

Reply:
Removing the AWS cloudwatch comment. Thanks

If a pod get ejected before flow finishes I don't think we still have logs in previous implementation.

Reply:
And if the Flow fails within?

Reply:
Manually tested RaiseErrorFlow and the exceptions are correctly captured: s3://serve-datalake-zillowgroup/zillow/workflow_sdk/metaflow_28d/dev/aip-example/RaiseErrorFlow/kfp-44b6944f-615e-49b0-9577-a485b500ea9a/error_step/kfp2/

Trying to automate this integration test now...